### PR TITLE
feat: sdk avatar modifier area

### PR DESCRIPTION
**Pull request by pravusjif**
`AvatarModifierArea` component support implementation.

* Refactored `CharacterTriggerArea` scripts to have better cleanup.
* Made sure `CameraModeArea` and `AvatarModifierArea` systems take care of resetting their effect if the component is removed or the entity is flagged for destruction.
* Implemented "ExcludeIds" feature for AvatarModifierArea including being able to dynamically change it and updating the state of affected avatars while they are still inside the area.
* SDK7 component usage docs: https://docs.decentraland.org/creator/development-guide/sdk7/avatar-modifiers/#placing-avatar-modifier-areas
* Tested with local scene with the following snippet:
```
const excludedUser = 'USER-PUBLIC-KEY-GOES-HERE'
const avatarModifierAreaEntity = engine.addEntity()
const avatarModifierAreaSize = Vector3.create(4, 4, 4)
Transform.create(avatarModifierAreaEntity, { 
position: Vector3.create(0, 1, 24),
scale: avatarModifierAreaSize
})
AvatarModifierArea.create(avatarModifierAreaEntity, {
area: avatarModifierAreaSize,
modifiers: [AvatarModifierType.AMT_HIDE_AVATARS],
excludeIds: []
})
MeshRenderer.setBox(avatarModifierAreaEntity)
Material.setPbrMaterial(avatarModifierAreaEntity, { albedoColor: Color4.create(0.5, 0.5, 0.5, 0.5) })
const avatarModifierAreaCube = createCube(1, 1, 27, false)
MeshCollider.setBox(avatarModifierAreaCube)
pointerEventsSystem.onPointerDown(
  { entity: avatarModifierAreaCube, opts: { button: InputAction.IA_POINTER, hoverText: 'toggle excluded id' } },
  () => {
    const mutable = AvatarModifierArea.getMutable(avatarModifierAreaEntity)
    if (mutable.excludeIds.includes(excludedUser)) {
      mutable.excludeIds = []
      console.log(`REMOVED excluded users`)
    }
    else {
      mutable.excludeIds = [excludedUser]
      console.log(`ADDED excluded users`)
    }
  }
)
```

**Implementation Notes**
*  “Hide Passport” mode: Not implementing since the whole passport features are needed for that.

https://github.com/decentraland/unity-explorer/assets/1031741/d4f5728e-66af-4b4e-9ca6-24132e122f26

### CODE REVIEW

**Review by popuz — APPROVED** [review]
LGTM 🟢

File: Explorer/Assets/Scripts/Global/StaticContainer.cs [hunk 1]
```diff
@@ -40,8 +40,8 @@ namespace Global
     public class StaticContainer : IDCLPlugin<StaticSettings>
     {
         private ProvidedInstance<CharacterObject> characterObject;
-        public readonly ObjectProxy<World> GlobalWorldProxy = new ();
-        public readonly ObjectProxy<AvatarBase> MainPlayerAvatarBaseProxy = new ();
+        public ObjectProxy<World> GlobalWorldProxy = new ();
+        public ObjectProxy<AvatarBase> MainPlayerAvatarBaseProxy = new ();
```
Comment (NickKhalow) [inline Explorer/Assets/Scripts/Global/StaticContainer.cs:44]: why to remove `readonly`?

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            ExcludedIds.Clear();
+
+            foreach (string excludedId in excludedIds)
+            {
+                // Profile component stores UserId in lowercase...
+                ExcludedIds.Add(excludedId.ToLower());
+            }
+        }
+    }
```
Comment (NickKhalow) [inline Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs:26]: RepeatedField can be replaced by IEnumerable

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+using Google.Protobuf.Collections;
+using System.Collections.Generic;
+
+namespace DCL.SDKComponents.AvatarModifierArea.Components
+{
+    public struct AvatarModifierAreaComponent
+    {
+        public readonly HashSet<string> ExcludedIds;
+
+        public AvatarModifierAreaComponent(RepeatedField<string> excludedIds)
```
Comment (NickKhalow) [inline Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs:10]: Here too RepeatedField

File: Explorer/Assets/DCL/AvatarRendering/AvatarShape/Systems/AvatarShapeVisibilitySystem.cs [hunk 4]
```diff
@@ -26,14 +26,25 @@ public override void Initialize()
...
+        private void UpdateAvatarsVisibilityState(ref AvatarShapeComponent avatarShape)
+        {
+            UpdateVisibilityState(ref avatarShape, avatarShape.HiddenByModifierArea);
+        }
 
+        private void UpdateVisibilityState(ref AvatarShapeComponent avatarShape, bool shouldBeHidden)
+        {
             switch (avatarShape.IsVisible)
             {
                 case true when shouldBeHidden:
```
Comment (NickKhalow) [inline Explorer/Assets/DCL/AvatarRendering/AvatarShape/Systems/AvatarShapeVisibilitySystem.cs:50]: minor: replace switch to if syntax, switch with booleans as cases seems a bit confusing

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Systems/AvatarModifierAreaHandlerSystem.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            foreach (Transform avatarTransform in triggerAreaComponent.CurrentAvatarsInside) { ToggleAvatarHiding(avatarTransform, false, modifierComponent.ExcludedIds); }
+        }
+
+        internal void ToggleAvatarHiding(Transform avatarTransform, bool shouldHide, HashSet<string> excludedIds)
+        {
+            var found = false;
+
+            // There's no way to do a Query/InlineQuery getting both entity and TransformComponent...
+            globalWorld.Query(in AVATAR_BASE_QUERY,
+                entity =>
```
Comment (NickKhalow) [inline Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Systems/AvatarModifierAreaHandlerSystem.cs:102]: Should be refactored without lambdas on every frame, it allocates much memory

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Systems/AvatarModifierAreaHandlerSystem.cs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+                    globalWorld.Get<AvatarShapeComponent>(entity).HiddenByModifierArea = shouldHide;
+                });
+        }
+
+        internal void CorrectAvatarHidingState(Transform avatarTransform, HashSet<string> excludedIds)
+        {
+            var found = false;
+
+            globalWorld.Query(in AVATAR_BASE_QUERY,
+                entity =>
```
Comment (NickKhalow) [inline Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Systems/AvatarModifierAreaHandlerSystem.cs:123]: same with lambdas

**Review by NickKhalow — CHANGES_REQUESTED** [review]
Good job, @pravusjif 

The major thing to rework those closures, because they are very memory consuming.
Also I left some comments what can be improved with method signatures.

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs (hunk 2 above)
Comment (NickKhalow) [reply Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs:26]: also with this ToLower transformation that are spreading around the code base let's bring here `struct UserId` that automatically translated it to lower case

**Review by sandrade-dcl — APPROVED** [review]
LGTM! good work!

File: Explorer/Assets/Scripts/Global/StaticContainer.cs (hunk 1 above)
Comment (pravusjif) [reply Explorer/Assets/Scripts/Global/StaticContainer.cs:44]: It was a problem when merging main, will add them back

Comment (pravusjif) [reply Explorer/Assets/Scripts/Global/StaticContainer.cs]: Tackled in [this commit](https://github.com/decentraland/unity-explorer/pull/336/commits/2ff94b5bfe636a43c73c3a8616c9bca7b55d70cb)

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs (hunk 2 above)
Comment (pravusjif) [reply Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs]: Tackled in [this commit](https://github.com/decentraland/unity-explorer/pull/336/commits/2ff94b5bfe636a43c73c3a8616c9bca7b55d70cb)

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs (hunk 3 above)
Comment (pravusjif) [reply Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs]: Tackled in [this commit](https://github.com/decentraland/unity-explorer/pull/336/commits/2ff94b5bfe636a43c73c3a8616c9bca7b55d70cb)

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs [hunk 7]
```diff
@@ -7,13 +7,13 @@ public struct AvatarModifierAreaComponent
...
 
-        public AvatarModifierAreaComponent(RepeatedField<string> excludedIds)
+        public AvatarModifierAreaComponent(IEnumerable excludedIds)
         {
             ExcludedIds = new HashSet<string>();
             SetExcludedIds(excludedIds);
         }
 
-        public void SetExcludedIds(RepeatedField<string> excludedIds)
+        public void SetExcludedIds(IEnumerable excludedIds)
```
Comment (NickKhalow) [inline Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs:16]: IEnumerable < string > ?

**Review by NickKhalow — CHANGES_REQUESTED** [review]

File: Explorer/Assets/Scripts/Global/StaticContainer.cs (hunk 1 above)
Comment (NickKhalow) [reply Explorer/Assets/Scripts/Global/StaticContainer.cs]: Good

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs (hunk 7 above)
Comment (NickKhalow) [reply Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs:16]: one minor change and it seems good to me

**Review by NickKhalow — APPROVED** [review]

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs (hunk 7 above)
Comment (pravusjif) [reply Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Components/AvatarModifierAreaComponent.cs]: Tackled [here](https://github.com/decentraland/unity-explorer/pull/336/commits/521720a02f3845b5a5f08dfaa2b72439c79c0793)

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Systems/AvatarModifierAreaHandlerSystem.cs (hunk 5 above)
Comment (pravusjif) [reply Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Systems/AvatarModifierAreaHandlerSystem.cs:102]: As agreed, I will create an issue to refactor that part!

File: Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Systems/AvatarModifierAreaHandlerSystem.cs (hunk 6 above)
Comment (pravusjif) [reply Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Systems/AvatarModifierAreaHandlerSystem.cs:123]: As agreed, I will create an issue to refactor that part!

Comment (pravusjif) [reply Explorer/Assets/DCL/SDKComponents/AvatarModifierArea/Systems/AvatarModifierAreaHandlerSystem.cs:123]: Issue created: https://github.com/decentraland/unity-explorer/issues/359

**Pull request merged**